src/auth/src/hash.rs

```diff
@@ -0,0 +1,227 @@
+        return Err(VerifyError::MalformedHash);
+    }
+    let auth_value = parts[2].split(':').collect::<Vec<&str>>();
+    if auth_value.len() != 2 {
+        return Err(VerifyError::MalformedHash);
+    }
+
+    let iterations = auth_info[0]
+        .parse::<u32>()
+        .map_err(|_| VerifyError::MalformedHash)?;
+
+    let salt = BASE64_STANDARD
+        .decode(auth_info[1])
+        .map_err(|_| VerifyError::MalformedHash)?;
+
+    let salt = salt.try_into().map_err(|_| VerifyError::MalformedHash)?;
+
+    Ok(HashOpts {
+        iterations: NonZeroU32::new(iterations).ok_or(VerifyError::MalformedHash)?,
+        salt,
+    })
+}
+
+/// The SCRAM-SHA-256 hash
+struct ScramSha256Hash {
+    /// The number of iterations used for hashing
+    iterations: NonZeroU32,
+    /// The salt used for hashing
+    salt: [u8; 32],
+    /// The server key
+    server_key: [u8; SHA256_OUTPUT_LEN],
+    /// The client key
+    client_key: [u8; SHA256_OUTPUT_LEN],
+}
+
+impl Display for ScramSha256Hash {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "SCRAM-SHA-256${}:{}${}:{}",
+            self.iterations,
+            BASE64_STANDARD.encode(&self.salt),
+            BASE64_STANDARD.encode(&self.client_key),
+            BASE64_STANDARD.encode(&self.server_key)
+        )
+    }
+}
+
+fn scram256_hash_inner(hashed_password: PasswordHash) -> ScramSha256Hash {
+    let signing_key = Key::new(HMAC_SHA256, &hashed_password.hash);
+    let client_key = hmac::sign(&signing_key, b"Client Key");
+    let server_key = hmac::sign(&signing_key, b"Server Key");
+
+    ScramSha256Hash {
+        iterations: hashed_password.iterations,
+        salt: hashed_password.salt,
+        server_key: server_key.as_ref().try_into().unwrap(),
+        client_key: client_key.as_ref().try_into().unwrap(),
+    }
+}
+
+fn hash_password_inner(opts: &HashOpts, password: &[u8]) -> [u8; SHA256_OUTPUT_LEN] {
+    let mut salted_password = [0u8; SHA256_OUTPUT_LEN];
+    pbkdf2::derive(
+        SHA256,
+        opts.iterations,
+        &opts.salt,
+        password,
+        &mut salted_password,
+    );
+    salted_password
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[mz_ore::test]
+    fn test_hash_password() {
+        let password = "password".to_string();
+        let hashed_password = hash_password(&password);
+        assert_eq!(hashed_password.iterations, DEFAULT_ITERATIONS);
+        assert_eq!(hashed_password.salt.len(), DEFAULT_SALT_SIZE);
+        assert_eq!(hashed_password.hash.len(), SHA256_OUTPUT_LEN);
+    }
+
+    #[mz_ore::test]
+    fn test_scram256_hash() {
+        let password = "password".to_string();
+        let scram_hash = scram256_hash(&password);
+
+        let res = scram256_verify(&password, &scram_hash);
+        assert!(res.is_ok());
+        let res = scram256_verify(&"wrong_password".to_string(), &scram_hash);
+        assert!(res.is_err());
+    }
```

Comment:
Turning this into a `proptest!` would be nice. Guards against any odd behavior with string handling, e.g. `$` or ` ` in the password itself